src/Features/CSharp/Portable/SignatureHelp/AbstractCSharpSignatureHelpProvider.cs

```diff
@@ -68,5 +96,287 @@ protected static SymbolDisplayPart NewLine()
+            if (arguments.Count != 0)
+            {
+                foreach (var method in methodGroup)
+                {
+                    if (FindParameterIndexIfCompatibleMethod(arguments, method, position, semanticModel, semanticFactsService, out var parameterIndex))
+                    {
+                        return (method, parameterIndex);
+                    }
+                }
+            }
+
+            // Note: Providing no recommendation if no arguments allows the model to keep the last implicit choice
+            return (null, -1);
+        }
+
+        /// <summary>
+        /// Simulates overload resolution with the arguments provided so far and determines if you might be calling this overload.
+        /// Returns true if an overload is acceptable. In that case, we output the parameter should be highlighted given the cursor's
+        /// position in the partial invocation.
+        /// </summary>
+        protected static bool FindParameterIndexIfCompatibleMethod(SeparatedSyntaxList<ArgumentSyntax> arguments, IMethodSymbol method, int position,
+            SemanticModel semanticModel, ISemanticFactsService semanticFactsService, out int foundParameterIndex)
+        {
+            // map the arguments to their corresponding parameters
+            var argToParamMap = PrepareArgToParamMap(arguments, method);
+            if (argToParamMap is null)
+            {
+                foundParameterIndex = -1;
+                return false;
+            }
+
+            // verify that the arguments are compatible with their corresponding parameters
+            var argumentCount = arguments.Count;
+            var parameters = method.Parameters;
+            for (var argumentIndex = 0; argumentIndex < argumentCount; argumentIndex++)
+            {
+                var parameterIndex = argToParamMap[argumentIndex];
+                if (parameterIndex < 0)
+                {
+                    continue;
+                }
+
+                var parameter = parameters[parameterIndex];
+                var argument = arguments[argumentIndex];
+
+                if (!IsCompatibleArgument(argument, parameter))
+                {
+                    foundParameterIndex = -1;
+                    argToParamMap.Free();
+                    return false;
+                }
+            }
+
+            // find the parameter at the cursor position
+            var argumentIndexToSave = TryGetArgumentIndex(arguments, position);
+            if (argumentIndexToSave >= 0)
+            {
+                var foundParam = argToParamMap[argumentIndexToSave];
+                if (foundParam >= 0)
+                {
+                    foundParameterIndex = foundParam;
+                }
+                else
+                {
+                    var firstUnspecified = FirstUnspecifiedParameter(argToParamMap, argumentCount);
+                    foundParameterIndex = firstUnspecified < 0 ? 0 : firstUnspecified;
+                }
```

Comment:
i feel like this needs to be doc'ed.

Reply:
Simplified now

Reply:
i still don't quite get it.  so you can return `foundParameterIndex  = -1` and `true` below?  is that intended/desirable?   i feel like i do need examples in the code which explain.